Mounting a previously searchable snapshotted index leads to extremely confusing errors about translog failure

I was manually testing (on master, b914994b40ec26bf8b593080f43bfc17d13b4068) a scenario it'd be nice to be able to use within ILM, and doing the following:

```
POST /_license/start_trial?acknowledge=true

PUT /_snapshot/repo1
{
  "type": "fs",
  "settings": {
    "location": "repo1"
  }
}

PUT /_snapshot/repo2
{
  "type": "fs",
  "settings": {
    "location": "repo2"
  }
}

POST /foo/_doc
{
  "foo": "bar"
}

PUT /_snapshot/repo1/snap1?wait_for_completion

POST /_snapshot/repo1/snap1/_mount?wait_for_completion
{
  "index": "foo",
  "renamed_index": "restored-foo",
  "index_settings": {
    "index.number_of_replicas": 0
  },
  "ignored_index_settings": [ "index.refresh_interval" ]
}

GET /restored-foo/_search

PUT /_snapshot/repo2/snap2?wait_for_completion

POST /_snapshot/repo2/snap2/_mount?wait_for_completion
{
  "index": "restored-foo",
  "renamed_index": "partial-restored-foo",
  "index_settings": {
    "index.number_of_replicas": 0
  },
  "ignored_index_settings": [ "index.refresh_interval" ]
}
```

Leads to restoring the `partial-restored-foo` index, but the index is red, and the ES log contains:

```
[2021-02-09T15:24:11,994][WARN ][o.e.i.CompositeIndexEventListener] [runTask-0] [partial-restored-foo] failed to invoke the listener before the shard recovery starts for [partial-restored-foo][0]
org.elasticsearch.index.translog.TranslogException: failed to associate a new translog
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.associateNewEmptyTranslogWithIndex(SearchableSnapshotIndexEventListener.java:107) ~[?:?]
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.beforeIndexShardRecovery(SearchableSnapshotIndexEventListener.java:68) ~[?:?]
	at org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:254) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1385) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:444) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:269) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1942) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2671) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:728) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: org.elasticsearch.index.store.ImmutableDirectoryException: file [corrupted_dR7BxgMnTAq9SfcYqbk0hg] is not mutable
	at org.elasticsearch.index.store.InMemoryNoOpCommitDirectory.ensureMutable(InMemoryNoOpCommitDirectory.java:126) ~[?:?]
	at org.elasticsearch.index.store.InMemoryNoOpCommitDirectory.createOutput(InMemoryNoOpCommitDirectory.java:79) ~[?:?]
	at org.apache.lucene.store.FilterDirectory.createOutput(FilterDirectory.java:74) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.elasticsearch.index.store.ByteSizeCachingDirectory.createOutput(ByteSizeCachingDirectory.java:118) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.apache.lucene.store.FilterDirectory.createOutput(FilterDirectory.java:74) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.elasticsearch.index.store.Store.markStoreCorrupted(Store.java:1354) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.store.Store.readLastCommittedSegmentsInfo(Store.java:181) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.associateNewEmptyTranslogWithIndex(SearchableSnapshotIndexEventListener.java:100) ~[?:?]
	... 13 more
[2021-02-09T15:24:12,006][WARN ][o.e.i.c.IndicesClusterStateService] [runTask-0] [partial-restored-foo][0] marking and sending shard failed due to [failed recovery]
org.elasticsearch.indices.recovery.RecoveryFailedException: [partial-restored-foo][0]: Recovery failed on {runTask-0}{-S4QWJmoQ7m0bDCUz-y4nQ}{aqye9hlGSn66hYEOqSDJRQ}{127.0.0.1}{127.0.0.1:9300}{cdfhilmrstw}{ml.machine_memory=67405721600, xpack.installed=true, transform.node=true, testattr=test, ml.max_open_jobs=20, ml.max_jvm_size=518979584}
	at org.elasticsearch.index.shard.IndexShard.lambda$executeRecovery$21(IndexShard.java:2724) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:125) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:274) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1942) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2671) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:728) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: org.elasticsearch.index.translog.TranslogException: failed to associate a new translog
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.associateNewEmptyTranslogWithIndex(SearchableSnapshotIndexEventListener.java:107) ~[?:?]
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.beforeIndexShardRecovery(SearchableSnapshotIndexEventListener.java:68) ~[?:?]
	at org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:254) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1385) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:444) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:269) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	... 8 more
Caused by: org.elasticsearch.index.store.ImmutableDirectoryException: file [corrupted_dR7BxgMnTAq9SfcYqbk0hg] is not mutable
	at org.elasticsearch.index.store.InMemoryNoOpCommitDirectory.ensureMutable(InMemoryNoOpCommitDirectory.java:126) ~[?:?]
	at org.elasticsearch.index.store.InMemoryNoOpCommitDirectory.createOutput(InMemoryNoOpCommitDirectory.java:79) ~[?:?]
	at org.apache.lucene.store.FilterDirectory.createOutput(FilterDirectory.java:74) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.elasticsearch.index.store.ByteSizeCachingDirectory.createOutput(ByteSizeCachingDirectory.java:118) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.apache.lucene.store.FilterDirectory.createOutput(FilterDirectory.java:74) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.elasticsearch.index.store.Store.markStoreCorrupted(Store.java:1354) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.store.Store.readLastCommittedSegmentsInfo(Store.java:181) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.associateNewEmptyTranslogWithIndex(SearchableSnapshotIndexEventListener.java:100) ~[?:?]
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.beforeIndexShardRecovery(SearchableSnapshotIndexEventListener.java:68) ~[?:?]
	at org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:254) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1385) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:444) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:269) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	... 8 more
[2021-02-09T15:24:12,018][WARN ][o.e.c.r.a.AllocationService] [runTask-0] failing shard [failed shard, shard [partial-restored-foo][0], node[-S4QWJmoQ7m0bDCUz-y4nQ], [P], recovery_source[snapshot recovery [o-d6zJ3_RPyfFPy6yzO7Pw] from repo2:snap2/9GwyHaIwQq2-OL8u2v72SQ], s[INITIALIZING], a[id=Wqh2UCJwS_uZIHZa_GM5zQ], unassigned_info[[reason=NEW_INDEX_RESTORED], at[2021-02-09T22:24:11.891Z], delayed=false, details[restore_source[repo2/snap2]], allocation_status[fetching_shard_data]], expected_shard_size[477], message [failed recovery], markAsStale [true], failure [org.elasticsearch.indices.recovery.RecoveryFailedException: [partial-restored-foo][0]: Recovery failed on {runTask-0}{-S4QWJmoQ7m0bDCUz-y4nQ}{aqye9hlGSn66hYEOqSDJRQ}{127.0.0.1}{127.0.0.1:9300}{cdfhilmrstw}{ml.machine_memory=67405721600, xpack.installed=true, transform.node=true, testattr=test, ml.max_open_jobs=20, ml.max_jvm_size=518979584}
	at org.elasticsearch.index.shard.IndexShard.lambda$executeRecovery$21(IndexShard.java:2724)
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:125)
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:274)
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1942)
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2671)
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:728)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: [partial-restored-foo/veKL0LHZS3WBR0o1L8Lsjw][[partial-restored-foo][0]] org.elasticsearch.index.translog.TranslogException: failed to associate a new translog
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.associateNewEmptyTranslogWithIndex(SearchableSnapshotIndexEventListener.java:107)
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.beforeIndexShardRecovery(SearchableSnapshotIndexEventListener.java:68)
	at org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:254)
	at org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1385)
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:444)
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:269)
	... 8 more
Caused by: org.elasticsearch.index.store.ImmutableDirectoryException: file [corrupted_dR7BxgMnTAq9SfcYqbk0hg] is not mutable
	at org.elasticsearch.index.store.InMemoryNoOpCommitDirectory.ensureMutable(InMemoryNoOpCommitDirectory.java:126)
	at org.elasticsearch.index.store.InMemoryNoOpCommitDirectory.createOutput(InMemoryNoOpCommitDirectory.java:79)
	at org.apache.lucene.store.FilterDirectory.createOutput(FilterDirectory.java:74)
	at org.elasticsearch.index.store.ByteSizeCachingDirectory.createOutput(ByteSizeCachingDirectory.java:118)
	at org.apache.lucene.store.FilterDirectory.createOutput(FilterDirectory.java:74)
	at org.elasticsearch.index.store.Store.markStoreCorrupted(Store.java:1354)
	at org.elasticsearch.index.store.Store.readLastCommittedSegmentsInfo(Store.java:181)
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.associateNewEmptyTranslogWithIndex(SearchableSnapshotIndexEventListener.java:100)
	... 13 more
]]
```

Especially concerning is the line:

```
org.elasticsearch.index.store.ImmutableDirectoryException: file [corrupted_dR7BxgMnTAq9SfcYqbk0hg] is not mutable
```

Since it named the file as "corrupted".

We lose the actual exception somewhere along the way; adjusting the logging a bit I get this:

```
[2021-02-10T08:54:48,954][WARN ][o.e.i.s.Store            ] [node-0] [partial-restored-foo][0] Can't mark store as corrupted
org.apache.lucene.index.CorruptIndexException: Unexpected file read error while reading index. (resource=BufferedChecksumIndexInput(ByteArrayIndexInput(segments_3)))
	at org.apache.lucene.index.SegmentInfos.readCommit(SegmentInfos.java:293) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.apache.lucene.index.SegmentInfos$1.doBody(SegmentInfos.java:461) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.apache.lucene.index.SegmentInfos$1.doBody(SegmentInfos.java:458) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.apache.lucene.index.SegmentInfos$FindSegmentsFile.run(SegmentInfos.java:720) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.apache.lucene.index.SegmentInfos$FindSegmentsFile.run(SegmentInfos.java:672) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.apache.lucene.index.SegmentInfos.readLatestCommit(SegmentInfos.java:463) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.elasticsearch.common.lucene.Lucene.readSegmentInfos(Lucene.java:132) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.store.Store.readSegmentsInfo(Store.java:194) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.store.Store.readLastCommittedSegmentsInfo(Store.java:179) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.associateNewEmptyTranslogWithIndex(SearchableSnapshotIndexEventListener.java:100) [searchable-snapshots-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotIndexEventListener.beforeIndexShardRecovery(SearchableSnapshotIndexEventListener.java:68) [searchable-snapshots-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:254) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1385) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:444) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:269) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1942) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2671) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:728) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
Caused by: java.io.FileNotFoundException: _0.si
	at org.elasticsearch.index.store.SearchableSnapshotDirectory.lambda$fileInfo$1(SearchableSnapshotDirectory.java:287) ~[?:?]
	at java.util.Optional.orElseThrow(Optional.java:401) ~[?:?]
	at org.elasticsearch.index.store.SearchableSnapshotDirectory.fileInfo(SearchableSnapshotDirectory.java:287) ~[?:?]
	at org.elasticsearch.index.store.SearchableSnapshotDirectory.openInput(SearchableSnapshotDirectory.java:381) ~[?:?]
	at org.elasticsearch.index.store.InMemoryNoOpCommitDirectory.openInput(InMemoryNoOpCommitDirectory.java:107) ~[?:?]
	at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:100) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:100) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.apache.lucene.store.Directory.openChecksumInput(Directory.java:157) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.apache.lucene.codecs.lucene86.Lucene86SegmentInfoFormat.read(Lucene86SegmentInfoFormat.java:91) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:08:18]
	at org.apache.lucene.index.SegmentInfos.readCommit(SegmentInfos.java:357) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	at org.apache.lucene.index.SegmentInfos.readCommit(SegmentInfos.java:291) ~[lucene-core-8.8.0.jar:8.8.0 b10659f0fc18b58b90929cfdadde94544d202c4a - noble - 2021-01-25 19:07:45]
	... 22 more
```

--- a/server/src/main/java/org/elasticsearch/index/store/Store.java
+++ b/server/src/main/java/org/elasticsearch/index/store/Store.java
@@ -1360,7 +1360,7 @@ public class Store extends AbstractIndexShardComponent implements Closeable, Ref
                 BytesRef ref = bytes.toBytesRef();
                 output.writeBytes(ref.bytes, ref.offset, ref.length);
                 CodecUtil.writeFooter(output);
-            } catch (IOException ex) {
+            } catch (IOException | ImmutableDirectoryException ex) {
                 logger.warn("Can't mark store as corrupted", ex);
             }
             directory().sync(Collections.singleton(corruptionMarkerName));

--- a/x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java
@@ -1306,7 +1306,7 @@ public class SearchableSnapshotsIntegTests extends BaseSearchableSnapshotsIntegT
             }
         });
         assertThat(remountException.getMessage(), allOf(
-                containsString("is a searchable snapshot backed by index"),
+                containsString("is a snapshot of a searchable snapshot index backed by index"),
                 containsString(repositoryName),
                 containsString(snapshotOne.getName()),
                 containsString(indexName),

--- a/x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java
@@ -9,6 +9,7 @@ package org.elasticsearch.xpack.searchablesnapshots;
 import com.carrotsearch.hppc.cursors.ObjectCursor;
 import org.apache.lucene.index.IndexFileNames;
 import org.apache.lucene.search.TotalHits;
+import org.elasticsearch.ExceptionsHelper;
 import org.elasticsearch.ResourceNotFoundException;
 import org.elasticsearch.action.admin.cluster.snapshots.restore.RestoreSnapshotResponse;
 import org.elasticsearch.action.admin.cluster.snapshots.status.SnapshotIndexShardStatus;
@@ -96,7 +97,9 @@ import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitC
 import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.getDataTiersPreference;
 import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotsConstants.SNAPSHOT_DIRECTORY_FACTORY_KEY;
 import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotsConstants.SNAPSHOT_RECOVERY_STATE_FACTORY_KEY;
+import static org.hamcrest.Matchers.allOf;
 import static org.hamcrest.Matchers.anyOf;
+import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.greaterThanOrEqualTo;
@@ -1288,6 +1291,30 @@ public class SearchableSnapshotsIntegTests extends BaseSearchableSnapshotsIntegT
         logger.info("--> finished restoring snapshot-2");
 
         assertTotalHits(restoredIndexName, originalAllHits, originalBarHits);
+
+        final IllegalArgumentException remountException = expectThrows(IllegalArgumentException.class, () -> {
+            try {
+                mountSnapshot(
+                        restoreRepositoryName,
+                        snapshotTwo.getName(),
+                        restoredIndexName,
+                        randomAlphaOfLength(10).toLowerCase(Locale.ROOT),
+                        Settings.EMPTY);
+            } catch (Exception e) {
+                final Throwable cause = ExceptionsHelper.unwrap(e, IllegalArgumentException.class);
+                throw cause == null ? e : cause;
+            }
+        });
+        assertThat(remountException.getMessage(), allOf(
+                containsString("is a searchable snapshot backed by index"),
+                containsString(repositoryName),
+                containsString(snapshotOne.getName()),
+                containsString(indexName),
+                containsString(restoreRepositoryName),
+                containsString(snapshotTwo.getName()),
+                containsString(restoredIndexName),
+                containsString("cannot be mounted; did you mean to restore it instead?")
+        ));
     }
 
     private void assertTotalHits(String indexName, TotalHits originalAllHits, TotalHits originalBarHits) throws Exception {

--- a/x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java
@@ -1295,17 +1295,20 @@ public class SearchableSnapshotsIntegTests extends BaseSearchableSnapshotsIntegT
         final IllegalArgumentException remountException = expectThrows(IllegalArgumentException.class, () -> {
             try {
                 mountSnapshot(
-                        restoreRepositoryName,
-                        snapshotTwo.getName(),
-                        restoredIndexName,
-                        randomAlphaOfLength(10).toLowerCase(Locale.ROOT),
-                        Settings.EMPTY);
+                    restoreRepositoryName,
+                    snapshotTwo.getName(),
+                    restoredIndexName,
+                    randomAlphaOfLength(10).toLowerCase(Locale.ROOT),
+                    Settings.EMPTY
+                );
             } catch (Exception e) {
                 final Throwable cause = ExceptionsHelper.unwrap(e, IllegalArgumentException.class);
                 throw cause == null ? e : cause;
             }
         });
-        assertThat(remountException.getMessage(), allOf(
+        assertThat(
+            remountException.getMessage(),
+            allOf(
                 containsString("is a snapshot of a searchable snapshot index backed by index"),
                 containsString(repositoryName),
                 containsString(snapshotOne.getName()),
@@ -1314,7 +1317,8 @@ public class SearchableSnapshotsIntegTests extends BaseSearchableSnapshotsIntegT
                 containsString(snapshotTwo.getName()),
                 containsString(restoredIndexName),
                 containsString("cannot be mounted; did you mean to restore it instead?")
-        ));
+            )
+        );
     }
 
     private void assertTotalHits(String indexName, TotalHits originalAllHits, TotalHits originalBarHits) throws Exception {

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
@@ -43,6 +43,7 @@ import org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots;
 import org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotsConstants;
 
 import java.util.Arrays;
+import java.util.Locale;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
@@ -178,6 +179,21 @@ public class TransportMountSearchableSnapshotAction extends TransportMasterNodeA
 
             final String[] ignoreIndexSettings = Arrays.copyOf(request.ignoreIndexSettings(), request.ignoreIndexSettings().length + 1);
             ignoreIndexSettings[ignoreIndexSettings.length - 1] = IndexMetadata.SETTING_DATA_PATH;
+
+            final IndexMetadata indexMetadata = repository.getSnapshotIndexMetaData(repoData, snapshotId, indexId);
+            if (INDEX_STORE_TYPE_SETTING.get(indexMetadata.getSettings()).equals(SearchableSnapshotsConstants.SNAPSHOT_DIRECTORY_FACTORY_KEY)) {
+                throw new IllegalArgumentException(String.format(Locale.ROOT,
+                        "index [%s] in snapshot [%s/%s:%s] is a searchable snapshot backed by index [%s] in snapshot [%s/%s:%s] " +
+                                "and cannot be mounted; did you mean to restore it instead?",
+                        indexName,
+                        repoName,
+                        repository.getMetadata().uuid(),
+                        snapName,
+                        SearchableSnapshots.SNAPSHOT_INDEX_NAME_SETTING.get(indexMetadata.getSettings()),
+                        SearchableSnapshots.SNAPSHOT_REPOSITORY_NAME_SETTING.get(indexMetadata.getSettings()),
+                        SearchableSnapshots.SNAPSHOT_REPOSITORY_UUID_SETTING.get(indexMetadata.getSettings()),
+                        SearchableSnapshots.SNAPSHOT_SNAPSHOT_NAME_SETTING.get(indexMetadata.getSettings())));
+            }
 
             client.admin()
                 .cluster()

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
@@ -51,6 +51,7 @@ import java.util.Optional;
 import static org.elasticsearch.index.IndexModule.INDEX_RECOVERY_TYPE_SETTING;
 import static org.elasticsearch.index.IndexModule.INDEX_STORE_TYPE_SETTING;
 import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.getDataTiersPreference;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotsConstants.isSearchableSnapshotStore;
 
 /**
  * Action that mounts a snapshot as a searchable snapshot, by converting the mount request into a restore request with specific settings
@@ -181,7 +182,7 @@ public class TransportMountSearchableSnapshotAction extends TransportMasterNodeA
             ignoreIndexSettings[ignoreIndexSettings.length - 1] = IndexMetadata.SETTING_DATA_PATH;
 
             final IndexMetadata indexMetadata = repository.getSnapshotIndexMetaData(repoData, snapshotId, indexId);
-            if (INDEX_STORE_TYPE_SETTING.get(indexMetadata.getSettings()).equals(SearchableSnapshotsConstants.SNAPSHOT_DIRECTORY_FACTORY_KEY)) {
+            if (isSearchableSnapshotStore(indexMetadata.getSettings())) {
                 throw new IllegalArgumentException(String.format(Locale.ROOT,
                         "index [%s] in snapshot [%s/%s:%s] is a searchable snapshot backed by index [%s] in snapshot [%s/%s:%s] " +
                                 "and cannot be mounted; did you mean to restore it instead?",

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
@@ -183,9 +183,11 @@ public class TransportMountSearchableSnapshotAction extends TransportMasterNodeA
 
             final IndexMetadata indexMetadata = repository.getSnapshotIndexMetaData(repoData, snapshotId, indexId);
             if (isSearchableSnapshotStore(indexMetadata.getSettings())) {
-                throw new IllegalArgumentException(String.format(Locale.ROOT,
-                        "index [%s] in snapshot [%s/%s:%s] is a snapshot of a searchable snapshot index " +
-                                "backed by index [%s] in snapshot [%s/%s:%s] and cannot be mounted; did you mean to restore it instead?",
+                throw new IllegalArgumentException(
+                    String.format(
+                        Locale.ROOT,
+                        "index [%s] in snapshot [%s/%s:%s] is a snapshot of a searchable snapshot index "
+                            + "backed by index [%s] in snapshot [%s/%s:%s] and cannot be mounted; did you mean to restore it instead?",
                         indexName,
                         repoName,
                         repository.getMetadata().uuid(),
@@ -193,7 +195,9 @@ public class TransportMountSearchableSnapshotAction extends TransportMasterNodeA
                         SearchableSnapshots.SNAPSHOT_INDEX_NAME_SETTING.get(indexMetadata.getSettings()),
                         SearchableSnapshots.SNAPSHOT_REPOSITORY_NAME_SETTING.get(indexMetadata.getSettings()),
                         SearchableSnapshots.SNAPSHOT_REPOSITORY_UUID_SETTING.get(indexMetadata.getSettings()),
-                        SearchableSnapshots.SNAPSHOT_SNAPSHOT_NAME_SETTING.get(indexMetadata.getSettings())));
+                        SearchableSnapshots.SNAPSHOT_SNAPSHOT_NAME_SETTING.get(indexMetadata.getSettings())
+                    )
+                );
             }
 
             client.admin()

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
@@ -184,8 +184,8 @@ public class TransportMountSearchableSnapshotAction extends TransportMasterNodeA
             final IndexMetadata indexMetadata = repository.getSnapshotIndexMetaData(repoData, snapshotId, indexId);
             if (isSearchableSnapshotStore(indexMetadata.getSettings())) {
                 throw new IllegalArgumentException(String.format(Locale.ROOT,
-                        "index [%s] in snapshot [%s/%s:%s] is a searchable snapshot backed by index [%s] in snapshot [%s/%s:%s] " +
-                                "and cannot be mounted; did you mean to restore it instead?",
+                        "index [%s] in snapshot [%s/%s:%s] is a snapshot of a searchable snapshot index " +
+                                "backed by index [%s] in snapshot [%s/%s:%s] and cannot be mounted; did you mean to restore it instead?",
                         indexName,
                         repoName,
                         repository.getMetadata().uuid(),

